synapse/handlers/room.py

```diff
@@ -1040,46 +1041,65 @@ async def _send_events_for_new_room(
         """
 
         creator_id = creator.user.to_string()
-
         event_keys = {"room_id": room_id, "sender": creator_id, "state_key": ""}
-
         depth = 1
         last_sent_event_id: Optional[str] = None
+        prev_event: List[str] = []
+        state_map = {}
+        auth_events = []
 
-        def create(etype: str, content: JsonDict, **kwargs: Any) -> JsonDict:
+        def create_event_dict(etype: str, content: JsonDict, **kwargs: Any) -> JsonDict:
             e = {"type": etype, "content": content}
 
             e.update(event_keys)
             e.update(kwargs)
 
             return e
 
-        async def send(etype: str, content: JsonDict, **kwargs: Any) -> int:
-            nonlocal last_sent_event_id
+        async def create_event(
+            etype: str,
+            content: JsonDict,
+            auth_event_ids: Optional[List[str]] = None,
+            **kwargs: Any,
+        ) -> EventBase:
             nonlocal depth
+            nonlocal prev_event
 
-            event = create(etype, content, **kwargs)
-            logger.debug("Sending %s in new room", etype)
-            # Allow these events to be sent even if the user is shadow-banned to
-            # allow the room creation to complete.
-            (
-                sent_event,
-                last_stream_id,
-            ) = await self.event_creation_handler.create_and_send_nonmember_event(
+            event_dict = create_event_dict(etype, content, **kwargs)
+
+            event = await self.event_creation_handler.create_event_for_batch(
                 creator,
-                event,
+                event_dict,
+                prev_event,
+                depth=depth,
+                auth_event_ids=auth_event_ids,
```

Comment:
I think we want to derive this from `state_map`? We can use `hs.get_event_auth_handler().compute_auth_events(..)` to go from state_map to auth_event_ids

Reply:
I am slightly confused by this-`compute_auth_events` takes an already-created event and calculates what type of events to pull out of the full state (the `state_map`) to auth that particular event, but in this case we need the auth events to create the event-I think I am missing something here?
My solution was to just collect and pass in the state events I knew would be needed for authorizing all the subsequent events (thus the create, membership and power level event ids become the `auth_event_ids` as they are created), but this might not be right?

Reply:
> takes an already-created event

You can give it an `EventBuilder` object, which doesn't need to have been finalised yet? You can get one from:

```
builder = self.event_builder_factory.for_room_version(
            room_version_obj, event_dict
        )
```

(You'd probably then want to change the new create_event method to take in `builder` rather than `event_dict`)

Reply:
Right this has been done.